### PR TITLE
osd: handle ceph specific config changes for the mclock scheduler

### PR DESCRIPTION
**Pull request by sunnyku**
The below ceph parameters are set automatically
while enabling the mclock scheduler with a built-in profile.
The user in this case will not be able to modify these
ceph specific config options during runtime.

  a. osd_async_recovery_min_cost
  b. osd_recovery_max_active{_hdd,_ssd}
  c. osd_max_backfills
  d. osd_recovery_sleep{_hdd,_ssd,_hybrid}

If the custom profile is enabled for the mclock scheduler,
the user can modify these parameters.

Signed-off-by: Sunny Kumar <sunkumar@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -9938,9 +9946,45 @@ void OSD::handle_conf_change(const ConfigProxy& conf,
...
+            "osd_recovery_max_active_ssd",stringify(recovery_max_active));
+      // Set high value for osd_max_backfill
+      cct->_conf.set_val("osd_max_backfills", stringify(1000));
+
+      // Disable recovery sleep
+      cct->_conf.set_val("osd_recovery_sleep", stringify(0));
+      cct->_conf.set_val("osd_recovery_sleep_hdd", stringify(0));
+      cct->_conf.set_val("osd_recovery_sleep_ssd", stringify(0));
+      cct->_conf.set_val("osd_recovery_sleep_hybrid", stringify(0));
+    }
```
Comment (jdurgin) [inline src/osd/OSD.cc:9983]: nit: else should be on same line as braces

**Review by jdurgin — APPROVED** [review]
nice and easy to read!

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -9938,9 +9946,45 @@ void OSD::handle_conf_change(const ConfigProxy& conf,
...
+      changed.count("osd_recovery_max_active_ssd") ||
+      changed.count("osd_async_recovery_min_cost")) {
+    if (cct->_conf.get_val<std::string>("osd_op_queue") == "mclock_scheduler" &&
+        cct->_conf.get_val<std::string>("osd_mclock_profile") != "custom") {
+      // Set ceph config option to meet QoS goals
+      // Set low recovery min cost
+      cct->_conf.set_val("osd_async_recovery_min_cost", stringify(1));
+      // Set high value for recovery max active
+      uint32_t recovery_max_active = 1000;
+      if (cct->_conf->osd_recovery_max_active)
```
Comment (sseshasa) [inline src/osd/OSD.cc:9966]: Add missing braces for the 'if' conditions here and other places. coding style is documented here: https://github.com/ceph/ceph/blob/master/CodingStyle

Comment (sseshasa) [inline src/osd/OSD.cc:9966]: It should be enough to set "osd_recovery_max_active" as it supersedes other options like osd_recovery_max_active_{ssd, hdd}.

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -9938,9 +9946,45 @@ void OSD::handle_conf_change(const ConfigProxy& conf,
...
+      changed.count("osd_recovery_sleep_hybrid") ||
+      changed.count("osd_recovery_max_active") ||
+      changed.count("osd_recovery_max_active_hdd") ||
+      changed.count("osd_recovery_max_active_ssd") ||
+      changed.count("osd_async_recovery_min_cost")) {
+    if (cct->_conf.get_val<std::string>("osd_op_queue") == "mclock_scheduler" &&
+        cct->_conf.get_val<std::string>("osd_mclock_profile") != "custom") {
+      // Set ceph config option to meet QoS goals
+      // Set low recovery min cost
+      cct->_conf.set_val("osd_async_recovery_min_cost", stringify(1));
```
Comment (tchaikov) [inline src/osd/OSD.cc:9963]: use `std::to_string()` if you please.

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -9938,9 +9946,45 @@ void OSD::handle_conf_change(const ConfigProxy& conf,
...
+        cct->_conf.set_val(
+            "osd_recovery_max_active", stringify(recovery_max_active));
+      if (store_is_rotational)
+        cct->_conf.set_val(
+            "osd_recovery_max_active_hdd", stringify(recovery_max_active));
+      else
+        cct->_conf.set_val(
+            "osd_recovery_max_active_ssd",stringify(recovery_max_active));
+      // Set high value for osd_max_backfill
+      cct->_conf.set_val("osd_max_backfills", stringify(1000));
```
Comment (tchaikov) [inline src/osd/OSD.cc:9976]: magic number here. is it required to be identical to `recovery_max_active`?

File: src/osd/OSD.cc [hunk 5]
```diff
@@ -9938,9 +9946,45 @@ void OSD::handle_conf_change(const ConfigProxy& conf,
...
+      uint32_t recovery_max_active = 1000;
+      if (cct->_conf->osd_recovery_max_active)
+        cct->_conf.set_val(
+            "osd_recovery_max_active", stringify(recovery_max_active));
+      if (store_is_rotational)
+        cct->_conf.set_val(
+            "osd_recovery_max_active_hdd", stringify(recovery_max_active));
+      else
+        cct->_conf.set_val(
+            "osd_recovery_max_active_ssd",stringify(recovery_max_active));
```
Comment (tchaikov) [inline src/osd/OSD.cc:9974]: please add a space after ",".

File: src/osd/OSD.cc (hunk 2 above)
Comment (tchaikov) [reply src/osd/OSD.cc:9966]: i concur with @sseshasa , would be great if you could add curly braces around the if branches. they helps to visualize the boundary of code block.

File: src/osd/OSD.cc (hunk 1 above)
Comment (sunnyku) [reply src/osd/OSD.cc]: done

File: src/osd/OSD.cc (hunk 2 above)
Comment (sunnyku) [reply src/osd/OSD.cc]: Done

File: src/osd/OSD.cc (hunk 3 above)
Comment (sunnyku) [reply src/osd/OSD.cc]: done

File: src/osd/OSD.cc (hunk 4 above)
Comment (sunnyku) [reply src/osd/OSD.cc]: Its not compulsory that the value for `osd_max_backfills` and `recovery_max_active` should be identical. 
However, test for mclock scheduler was conducted with these specific (high) values and the results were conclusive.

**Review by jdurgin — APPROVED** [review]

File: src/osd/OSD.cc [hunk 6]
```diff
@@ -9910,6 +9910,14 @@ const char** OSD::get_tracked_conf_keys() const
...
     "osd_pg_epoch_max_lag_factor",
     "osd_pg_epoch_persisted_max_stale",
+    "osd_recovery_sleep",
+    "osd_recovery_sleep_hdd",
+    "osd_recovery_sleep_ssd",
+    "osd_recovery_sleep_hybrid",
+    "osd_recovery_max_active",
+    "osd_recovery_max_active_hdd",
+    "osd_recovery_max_active_ssd",
+    "osd_async_recovery_min_cost"
```
Comment (neha-ojha) [inline src/osd/OSD.cc:9920]:
```suggestion
    "osd_async_recovery_min_cost",
```

**Review by neha-ojha — APPROVED** [review]

File: src/osd/OSD.cc (hunk 6 above)
Comment (sunnyku) [reply src/osd/OSD.cc]: Done, thanks for pointing out.

**Comment by sunnyku** [comment]
jenkins test api

**Comment by sunnyku** [comment]
jenkins retest this please

**Review by sseshasa — APPROVED** [review]
LGTM

**Comment by neha-ojha** [comment]
https://pulpito.ceph.com/nojha-2021-01-16_20:47:46-rados-wip-config-sets-ci-distro-basic-smithi/ - https://tracker.ceph.com/issues/48906 is a known issue and will be addressed in a follow-on PR.

**Comment by smithfarm** [comment]
@jdurgin @neha-ojha I noticed that this PR (which was merged on January 17) is in the pacific branch. Indeed, the pacific branch is identical to master even today. Wasn't it supposed to be split off on January 15?

**Comment by smithfarm** [comment]
> @jdurgin @neha-ojha I noticed that this PR (which was merged on January 17) is in the pacific branch. Indeed, the pacific branch is identical to master even today. Wasn't it supposed to be split off on January 15?

Well, nevermind I guess - after reading https://docs.ceph.com/en/latest/dev/developer_guide/merging/ I realized that cherry-picking to pacific cannot start until Phase 2 - https://docs.ceph.com/en/latest/dev/developer_guide/merging/#stable-release-candidates-i-e-x-1-z-phase-2 - starts, and for that to happen pacific needs a 16.1.x tag which it doesn't have yet.

**Pull request merged**